src/controller/model/endpoint.ts

```diff
@@ -261,43 +264,62 @@ class Endpoint extends Entity {
         return this.pendingRequests.length > 0;
     }
 
-    public async sendPendingRequests(): Promise<void> {
-        while (this.pendingRequests.length > 0) {
-            const r = this.pendingRequests.shift();
-            try {
-                const result = await r.func();
-                r.resolve(result);
-            } catch (error) {
-                r.reject(error);
+    public async sendPendingRequests(fastPolling: boolean): Promise<void> {
+        const handled: PendingRequest[] = [];
+        for (const request of this.pendingRequests) {
+            if (fastPolling || request.queueSendWhen == 'active') {
+                try {
+                    const result = await request.func();
+                    request.resolve(result);
+                } catch (error) {
+                    request.reject(error);
+                }
+                handled.push(request);
             }
         }
+
+        this.pendingRequests = this.pendingRequests.filter((r) => !handled.includes(r));
     }
 
-    private async queueRequest<Type>(func: () => Promise<Type>): Promise<Type> {
+    private async queueRequest<Type>(func: () => Promise<Type>, queueSendWhen: 'active' | 'fastpoll'): Promise<Type> {
         debug.info(`Sending to ${this.deviceIeeeAddress}/${this.ID} when active`);
         return new Promise((resolve, reject): void =>  {
-            this.pendingRequests.push({func, resolve, reject});
+            this.pendingRequests.push({func, resolve, reject, queueSendWhen});
         });
     }
 
-    private async sendRequest<Type>(func: () => Promise<Type>, sendWhenActive: boolean): Promise<Type> {
+    private async sendRequest<Type>(func: () => Promise<Type>, 
+        sendWhen: 'active' | 'auto' | 'immediate'): Promise<Type> {
+        const device = this.getDevice();
+        const fastPollingAvailable = !device.useImplicitCheckin;
+
+        let queueSendWhen: 'fastpoll' | 'immediate' | 'active';
+        if (sendWhen === 'auto') {
+            if (fastPollingAvailable) queueSendWhen = 'fastpoll';
+            else if (device.defaultSendWhenActive) queueSendWhen = 'active';
+            else queueSendWhen = 'immediate';
+        } else {
+            queueSendWhen = sendWhen;
+        }
+
         // If we already have something queued, we queue directly to avoid
         // messing up the ordering too much.
-        if (sendWhenActive && this.pendingRequests.length > 0) {
-            return this.queueRequest(func);
+        if (queueSendWhen !== 'immediate' && this.pendingRequests.length > 0) {
+            return this.queueRequest(func, queueSendWhen);
         }
 
+        // TODO
         try {
             return await func();
         } catch(error) {
-            if (!sendWhenActive) {
+            if (queueSendWhen === 'immediate') {
                 throw(error);
             }
         }
```

Comment:
> [this] still feels a bit strange, even when sendWhen == active or fastpoll the request will be send immediately, do you remember the reason for this?

It is strange, and it's a little annoying to effectively have retry logic both at this level and at adapter level...

The reason for attempting to send immediately was twofold:
1. To minimize responsiveness impact when sendWhenActive suddenly got enabled device-wide, as it previously only affected 4 individual and highly specific messages.
2. To improve sendWhenActive behavior when sending in response to a received message which was otherwise postponed till next time the queue flushed. An example would be updated load estimate from danfoss thermostats, which immediately result in sending the mean load.

The behavior was implemented in https://github.com/Koenkk/zigbee-herdsman/pull/447. Previous implementations of this used a short timeout after the queue emptied during which instantaneous sending was allowed. Maybe going back to something like that would be prettier, and when fastpoll is in use we can control exactly how long such a timeout should be.

Reply:
Makes sense, lets keep it.